url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/triggers/CronExpression.java

```diff
@@ -0,0 +1,1571 @@
+        }
+
+        StringTokenizer vTok = new StringTokenizer(expr, ",");
+        while (vTok.hasMoreTokens()) {
+          String v = vTok.nextToken();
+          storeExpressionVals(0, v, exprOn);
+        }
+
+        exprOn++;
+      }
+
+      if (exprOn <= DAY_OF_WEEK) {
+        throw new ParseException("Unexpected end of expression.", expression.length());
+      }
+
+      if (exprOn <= YEAR) {
+        storeExpressionVals(0, "*", YEAR);
+      }
+
+      TreeSet<Integer> dow = getSet(DAY_OF_WEEK);
+      TreeSet<Integer> dom = getSet(DAY_OF_MONTH);
+
+      // Copying the logic from the UnsupportedOperationException below
+      boolean dayOfMSpec = !dom.contains(NO_SPEC);
+      boolean dayOfWSpec = !dow.contains(NO_SPEC);
+
+      if (!dayOfMSpec || dayOfWSpec) {
+        if (!dayOfWSpec || dayOfMSpec) {
+          throw new ParseException("Support for specifying both a day-of-week AND a day-of-month parameter is not implemented.", 0);
+        }
+      }
+    } catch (ParseException pe) {
+      throw pe;
+    } catch (Exception e) {
+      throw new ParseException("Illegal cron expression format (" + e.toString() + ")", 0);
+    }
+  }
+
+  protected int storeExpressionVals(int pos, String s, int type) throws ParseException {
+
+    int incr = 0;
+    int i = skipWhiteSpace(pos, s);
+    if (i >= s.length()) {
+      return i;
+    }
+    char c = s.charAt(i);
+    if ((c >= 'A') && (c <= 'Z') && (!s.equals("L")) && (!s.equals("LW")) && (!s.matches("^L-[0-9]*[W]?"))) {
+      String sub = s.substring(i, i + 3);
+      int sval = -1;
+      int eval = -1;
+      if (type == MONTH) {
+        sval = getMonthNumber(sub) + 1;
+        if (sval <= 0) {
+          throw new ParseException("Invalid Month value: '" + sub + "'", i);
+        }
+        if (s.length() > i + 3) {
+          c = s.charAt(i + 3);
+          if (c == '-') {
+            i += 4;
+            sub = s.substring(i, i + 3);
+            eval = getMonthNumber(sub) + 1;
+            if (eval <= 0) {
+              throw new ParseException("Invalid Month value: '" + sub + "'", i);
+            }
+          }
+        }
+      } else if (type == DAY_OF_WEEK) {
+        sval = getDayOfWeekNumber(sub);
+        if (sval < 0) {
+          throw new ParseException("Invalid Day-of-Week value: '" + sub + "'", i);
+        }
+        if (s.length() > i + 3) {
+          c = s.charAt(i + 3);
+          if (c == '-') {
+            i += 4;
+            sub = s.substring(i, i + 3);
+            eval = getDayOfWeekNumber(sub);
+            if (eval < 0) {
+              throw new ParseException("Invalid Day-of-Week value: '" + sub + "'", i);
+            }
+          } else if (c == '#') {
+            try {
+              i += 4;
+              nthdayOfWeek = Integer.parseInt(s.substring(i));
+              if (nthdayOfWeek < 1 || nthdayOfWeek > 5) {
+                throw new Exception();
+              }
+            } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)